### ISSUE TITLE
Reading TFRecord through tf.data extremely slow in comparison to Pandas (MWE)

### ISSUE THREAD

**Issue by mimxrt** [tensorflow/tensorflow#37093]
**System information** 
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04.3 LTS
- TensorFlow installed from (source or binary): binary pip
- TensorFlow version (use command below): v2.1.0-rc2-17-ge5bf8de 2.1.0
- Python version: Python 3.7.6
- CUDA/cuDNN version: 7.6.4.38-1
- GPU model and memory: Quadro RTX 6000 22752 MB

**Describe the current behavior**

In the example below, first `tf.Data` is used to read data from dummy `TFRecord` files. Then pandas is used to read the same data from serialized `pandas.DataFrame` pickle-files. Both pipelines read chunks of `batch_size=512` elements from the data to the end (in the example, 10 files of 210083x135 floats).

In comparison, the performance of the `tf.Dataset` is very slow. Pandas reads the data  in ~**1.86 s** (~4.52 MB/s) vs `tf.data` in ~**51.3 s** (~0.16 MB/s).

The pipeline (in my understanding) does nothing more than to read the files from disk into batched junks in memory. Please help me understand and resolve this performance gap. 

This issue induces a huge performance hit in my model training, which uses a slightly more complicated pipeline.

**Describe the expected behavior**

Processing TFRecord files should be equally fast if not faster than processing pickle-files with pandas.

**Standalone code to reproduce the issue** 

```
import math
import time
import pandas
import random
import numpy as np
import tensorflow as tf

print("TensorFlow {}".format(tf.__version__))

num_features = 130
num_labels = 5
batch_size = 512
win_size = 2048
len_max = 210083

#########################################
# dummy data
#########################################

l1 = len_max
X1, Y1 = (np.ones((l1, num_features), dtype=np.float32), np.zeros((l1, num_labels), dtype=np.float32))

k = 10

#########################################
# create dummy tfrecord file(s)
#########################################

def write_tfrecord(X, Y, fn):
    with tf.Graph().as_default():
        ds = tf.data.Dataset.from_tensor_slices((X, Y))

        sstring = ds.map(lambda *x: 
           tf.reshape(tf.py_function(lambda *v:
               tf.train.Example(features=tf.train.Features(feature={
                   "features": tf.train.Feature(float_list=tf.train.FloatList(value=v[0].numpy())),
                   "label": tf.train.Feature(float_list=tf.train.FloatList(value=v[1].numpy())),
               })).SerializeToString(), x, tf.string
           ), ())
        )

        writer = tf.data.experimental.TFRecordWriter(fn)
        writer_op = writer.write(sstring)

        sess = tf.compat.v1.Session()
        sess.run(tf.compat.v1.global_variables_initializer())
        sess.run(writer_op)
        sess.close()

files_base_tf = ["./temp3.tfrecord"]

write_tfrecord(X1, Y1, files_base_tf[0])

#########################################
# create dummy pandas pickle file(s)
#########################################

df = pandas.DataFrame(np.concatenate((X1, Y1), axis=1))

files_base_pd = ["./temp3.pkl"]

df.to_pickle(files_base_pd[0])

#########################################
# take k random files as tf.Dataset
#########################################

_files = random.choices(files_base_tf, k=k)

ds_fs = tf.data.Dataset.list_files(_files, shuffle=False, seed=1)
fs_len = 0
for f in ds_fs:
    fs_len += 1
print("Reading {} tfrecord files...".format(fs_len))

# prepare the tf.Dataset

def prep_ds_file(file):
    _ds = tf.data.TFRecordDataset(file)
    _ds = _ds.map(lambda x: tf.io.parse_single_example(x, {
        "features": tf.io.FixedLenFeature([num_features], tf.float32),
        "label": tf.io.FixedLenFeature([num_labels], tf.float32),
    #}), num_parallel_calls=tf.data.experimental.AUTOTUNE)
    }), num_parallel_calls=1)
    #print(_ds)
    
    _ds = _ds.flat_map(lambda v: tf.data.Dataset.from_tensors((v["features"], v["label"])))
    #print(_ds)

    _ds = _ds.batch(batch_size, drop_remainder=False)
    print(_ds)
    
    return _ds


def prep_ds(fs):
    _ds = fs.interleave(prep_ds_file, num_parallel_calls=tf.data.experimental.AUTOTUNE)
    print(_ds)

    return _ds


ds = prep_ds(ds_fs)

#########################################
# read/use the tf.Dataset
#########################################

shapes_tf = []

ts = time.perf_counter()
i = 0
for e in ds:
    v = (lambda: [x.numpy() for x in e])()
    #shapes_tf.append([x.shape for x in v])
    i += 1
te = time.perf_counter()
print("Duration TFRecord ({} rounds): {} s".format(i, (te - ts)))
print("Speed: {} MB/s".format((k*l1*4/1000/1000/(te - ts))))

#########################################
# take k random files as pandas DataFrame
#########################################

_files = random.choices(files_base_pd, k=k)

#########################################
# read/use the pandas DataFrame
#########################################

shapes_pd = []

ts = time.perf_counter()
i = 0
for f in _files:
    df = pandas.read_pickle(f)
    n = math.ceil(len(df.index) / batch_size)
    for b in range(n):
        #print(df.iloc[(b*batch_size):((b+1)*batch_size)].shape)
        v = (lambda: df.iloc[(b*batch_size):((b+1)*batch_size)].to_numpy())()
        #shapes_pd.append(v.shape)
        i += 1
te = time.perf_counter()
print("Duration Pandas ({} rounds): {} s".format(i, (te - ts)))
print("Speed: {} MB/s".format((k*l1*4/1000/1000/(te - ts))))
```

**Other info / logs**

Output:
```
TensorFlow 2.1.0
Reading 10 tfrecord files...
<BatchDataset shapes: ((None, 130), (None, 5)), types: (tf.float32, tf.float32)>
<ParallelInterleaveDataset shapes: ((None, 130), (None, 5)), types: (tf.float32, tf.float32)>
Duration TFRecord (4110 rounds): 51.2759778869804 s
Speed: 0.1638841489970629 MB/s
Duration Pandas (4110 rounds): 1.8584379029925913 s
Speed: 4.521711479554074 MB/s
```

**Comment by amahendrakar** [comment]
Was able to reproduce the issue. Please find the Gist [here](https://colab.sandbox.google.com/gist/amahendrakar/983e36d410aa52fedaab9dc19135019c/37093.ipynb). Thanks!

**Comment by suphoff** [comment]
The overhead of parsing an example record is rather high compared to copying the small features and label fields. Changing the dataset to just passing the records speeds up the reads to 1/2 of the speed of the pandas.read_pick based solution. (Using @amahendrakar 's colab as a base )
This seems reasonable as reading dynamic sized records, calculating the record checksum, .. takes effort.
For larger feature and label tensors the example record parsing overhead should be a smaller percentage.
If example records are required - maybe storing batches of features/labels inside of a single example record (and unbatching them on read) could be a workaround for small tensors?
( I would also recommend looking at tf.data.experimental.ThreadingOptions if not all cores are busy )

**Comment by mimxrt** [comment]
Thank you for your reply. I have two questions if you allow:

1. I don't know if example records are required. Is there a better way to represent the data given in my MWE? If so, could you provide a code snippet or web link illustrating this?

2. Could you additionally or alternatively present a code snippet or point me to an example that shows how to properly use batches during storing/retrieving? It would help a lot.

Thanks again!

**Comment by mimxrt** [comment]
I found that using `FixedLenSequenceFeature` and storing/reading whole columns instead of rows mitigates the issue. It requires to use `ds = ds.map(lambda x : tf.transpose(x, [1, 0]))` after reading but works much faster. I guess that is what you meant? Anyway, thanks, I am closing this issue.

**Comment by marwan116** [comment]
@mimxrt - thank you for raising this issue and for mentioning the way you resolved it. Is there any chance you can post the code of the solution that makes use of `FixedLenSequenceFeature`? would be quite grateful

**Comment by mimxrt** [comment]
@marwan116 It has been a long time but I think this was the code snippet that mitigated this issue:

```
def prep_ds_file(file):
    _ds = tf.data.TFRecordDataset(
        file,
        #num_parallel_reads=tf.data.experimental.AUTOTUNE
    )
    
#     _ds = _ds.map(lambda x: tf.io.parse_single_example(x, {
#         "features": tf.io.FixedLenFeature([num_features], tf.float32),
#         "label": tf.io.FixedLenFeature([num_labels], tf.float32),
#     }), num_parallel_calls=tf.data.experimental.AUTOTUNE)
#     #}), num_parallel_calls=1)
#     #print(_ds)

    _ds = _ds.batch(l1)
    print(_ds)
    _ds = _ds.map(
        lambda x: tf.io.parse_example(x, {
            "features": tf.io.FixedLenFeature([num_features], tf.float32),
            "label": tf.io.FixedLenFeature([num_labels], tf.float32),
        }),
        #num_parallel_calls=tf.data.experimental.AUTOTUNE
    )
    print(_ds)
    
    #_ds = _ds.flat_map(lambda v: tf.data.Dataset.from_tensors((v["features"], v["label"])))
    _ds = _ds.interleave(
        lambda v: tf.data.Dataset.from_tensors((v["features"], v["label"])),
        #num_parallel_calls=tf.data.experimental.AUTOTUNE
    )
    print(_ds)

    _ds = _ds.unbatch()
    
    _ds = _ds.batch(batch_size, drop_remainder=False)
    print(_ds)
    
    return _ds


def prep_ds(fs):
#     _ds = fs.interleave(
#         prep_ds_file,
#         #num_parallel_calls=tf.data.experimental.AUTOTUNE
#     )
    _ds = fs.flat_map(prep_ds_file)
    
    _ds = _ds.prefetch(buffer_size=tf.data.experimental.AUTOTUNE)
    
    return _ds


ds = prep_ds(ds_fs)
```

Note the commented-out parts that were replaced from the original MWE. IIRC the solution was to read bigger chunks at once instead of "line by line" (using `parse_example` instead of `parse_single_example` with the concept _batch_ -> _read_ -> _unbatch_). Hope that helps!

**Comment by marwan116** [comment]
@mimxrt thank you for following up on this - I can see how employing `parse_example` to read a chunk of examples at a time will help optimize the read operation. 

But If I understand it correctly, this is different from your last comment on writing and reading the full columns which is the approach that I think would be the fastest for reading and writing

**Comment by mimxrt** [comment]
I don't exactly remember what I meant by that but I double checked and could not find any other example in connection to this issue. However, the way I ended up creating and using these datasets is as follows:

```
def _write(df, fn):
    # df is a Pandas DataFrame, fn is the filename

    def tf_serialize_example(*x):
        tf_string = tf.py_function(
            lambda *v : tf.train.Example(features=tf.train.Features(
                feature = {
                    l: tf.train.Feature(float_list=tf.train.FloatList(value=f)) for l, f in zip(df.columns, v)
                }
            )).SerializeToString(),
            x,
            tf.string
        )
        return tf.reshape(tf_string, ())

    ds = tf.data.Dataset.from_tensor_slices(
        tuple(df[col].values.reshape(1, -1) for col in df.columns)
    )

    ds_ser = ds.map(tf_serialize_example)

    writer = tf.data.experimental.TFRecordWriter(fn)
    writer.write(ds_ser)
```

For reading this, I actually end up using `parse_single_example`:

```
for i, filename in enumerate(filenames):
    path = my_path + filename
    ds_fs = tf.data.TFRecordDataset(
        path,
        num_parallel_reads=tf.data.experimental.AUTOTUNE,
    )
    #print(ds_fs)

    # add or remove features to **read** from the TFRecord here
    ds = ds_fs.map(lambda x : tf.io.parse_single_example(
        x, features={
#             "Ft1": tf.io.FixedLenSequenceFeature([], tf.float32, allow_missing=True),
#             **{str(i) : tf.io.FixedLenSequenceFeature([], tf.float32, allow_missing=True) for i in range(num_features-3)},
#             "Ft2": tf.io.FixedLenSequenceFeature([], tf.float32, allow_missing=True),
#             "Ft3": tf.io.FixedLenSequenceFeature([], tf.float32, allow_missing=True),
            "Label": tf.io.FixedLenSequenceFeature([], tf.float32, allow_missing=True),
        }
    ))

    # add or remove features to **include** in the Dataset here
    ds = ds.interleave(
        lambda v: tf.data.Dataset.from_tensors([v["Label"]]),
        #num_parallel_calls=tf.data.experimental.AUTOTUNE
    )
    #print(ds)
```
